### DISCUSSION TITLE
Can you recursively call derived atoms?

### DISCUSSION THREAD

**Discussion by ASgmbv** [pmndrs/jotai#1504] [Q&A]
Hi everyone, I have a following case:

```
export const addElementAtom = atom(
   null,
   (get, set, payload: { type: string; }
) => {

   // some code

   if (type === Section) {
     set(addElementAtom,
       {
         type: 'Box'
       }
     )
   }

   // some code

  }
)

```
How can I achieve this? It seems like it is not working for now

**Comment by dai-shi** [comment]
Use two atoms: base atom and derived atom.

`atomWithReducer` should also help this case.

**Reply by ASgmbv** [reply]
I have the base atom

```
const elements = atom<string[]>([]);

export const addElementAtom = atom(
   null,
   (get, set, payload: { type: string; }
) => {

    set(withImmer(elementsAtom), (prev) => {
      prev.push({
        type
      })
    });

   if (type === Section) {
     set(addElementAtom,
       {
         type: 'Box'
       }
     )
   }

  }
)
```

**Reply by dai-shi** [reply]
```js
    set(withImmer(elementsAtom), (prev) => {
      prev.push({
        type
      })
    });
```
looks interesting. while that may work, it might be better to use `produce`:
```js
    set(elementsAtom, produce((prev) => {
      prev.push({
        type
      })
    }));
```

**Comment by dai-shi** [comment]
> Hi everyone, I have a following case:
> 
> ```
> export const addElementAtom = atom(
>    null,
>    (get, set, payload: { type: string; }
> ) => {
> 
>    // some code
> 
>    if (type === Section) {
>      set(addElementAtom,
>        {
>          type: 'Box'
>        }
>      )
>    }
> 
>    // some code
> 
>   }
> )
> ```
> 
> How can I achieve this? It seems like it is not working for now

It would work in JS. Is your question about TS error?
